### ISSUE TITLE
Some Concerning Client Backend Implementation Details

### ISSUE THREAD

**Issue by 13-05** [tungdo0602/Discord-Web-Client#1]
## Concerns With Implementation
Hey! This seems like a really cool concept (I've only read through the source, not tried it yet :p), but there are some underlying concerns I have about this, and details others may want to be aware of if not changed. I'll try to be detailed:

### Your `IDENTIFY`
Line(s) of concern: [`payload`](https://github.com/tungdo0602/Discord-Web-Client/blob/main/libs/main.js#L78)

Alright, this is my first source of concern.
It seems your `"IDENTIFY"` payload is quite different from the client's (maybe you got inspiration from [this dude's response to one of my stackoverflow questions?](https://stackoverflow.com/a/71362558/17055513))

Regardless, here's an example of how it's *supposed to look* (source: adapted `"IDENTIFY"` payload from <https://github.com/dolfies/discord.py-self/blob/master/discord/gateway.py#L428> and <https://github.com/Merubokkusu/Discord-S.C.U.M/blob/master/discum/gateway/gateway.py#L83>, AKA the most-safe and most-functional selfbot libraries in my opinion):

```js
{
    "token": "TOKEN",
    "capabilities": 509,
    "properties": "SUPER-PROPERTIES", // should mirror x-super-properties iirc
    "presence": {
        "status": "online",
        "since": 0,
        "activities": [],
        "afk": false
    },
    "compress": false,
    "client_state": {
        "guild_hashes": {},
        "highest_last_message_id": "0",
        "read_state_version": 0,
        "user_guild_settings_version": -1,
        "user_settings_version": -1
    }
}
```

### Your `sendMessage` and `getMessage` Functions
Line(s) of concern: [`sendMessage`](https://github.com/tungdo0602/Discord-Web-Client/blob/main/libs/main.js#L212), [`getMessage`](https://github.com/tungdo0602/Discord-Web-Client/blob/main/libs/main.js#L215)

I understand wanting to "write everything from scratch!", but maybe you'd benefit from copying [Discord's internal functions](https://github.com/13-05/hidden-disc-docs#discords-api-frontend-a-custom-webpack-implementation)? After all, Discord would never detect requests exactly the same as theirs. If you're too lazy to go through and deobfuscate the source of their functions, don't worry, that's me too, then maybe you could write a simple import system for Discord's webpack modules and call them directly from Discord's own webpack, like ~~PoweRcord~~ [Powercord](https://powercord.dev/) does? Though, unless you call more functions besides these two, it'd somewhat bloat your webapp more than necessary.

Check out the Discord Datamining team's decompression of `window.webpackChunkdiscord_app` and the like [here](https://raw.githubusercontent.com/Discord-Datamining/Discord-Datamining/master/current.js), and run `CTRL` + `F` on the page to see if you can find the source for any function you'd like to copy if you don't like the idea of a webpack importation system.

### CORS Bypassing
Line(s) of concern: `N/A`

So, I've read through the source and it's pretty good, although how do you plan on bypassing CORS? Because, if you don't know: browsers "yell at you" if you make requests to a URL that has no `Access-Control-Allow-Origin` header (and one that allows you to make calls from other windows, at that) present on the response.

If you do plan on making this full-scale, maybe a browser extension would work? Not too sure, never messed with that.

### The Bot Token Functionality
Line(s) of concern: `N/A`

Alright, this is one idea I really like, so I won't try to nitpick *too much*. *Good* bot clients have been quite scarce ever since [Flam3rBoy's client](https://github.com/Flam3rboy/discord-bot-client) was discontinued, so another in the game would be awesome, especially if it's a web one!

But, there's always an issue. I'm not sure if you're aware of this, but Discord blocks all requests to a `Bot` account if a `User-Agent` is present on the headers. So, a couple ways to bypass this would be to
> a) Write a proxy server that removes the `User-Agent` header from requests, then routes the modified request to Discord
> b) Write a browser extension that removes the `User-Agent` header on requests.

Depending on your JS experience (*I* don't have much JS experience, so I'd lean toward proxies written in another language haha) you may find one easier or more difficult than the other. All the same, this should be an apparent issue.

### Your Sockets
Line(s) of concern: [`WebSocket`](https://github.com/tungdo0602/Discord-Web-Client/blob/main/libs/main.js#L76)

It seems that you're connecting to a [now-decommissioned API version](https://github.com/discord/discord-api-docs/discussions/4657). And actually, it was decommissioned *today*! What a time for me to stumble unto your client, haha. Anyway, here are the basics:
- Outdated WebSocket URL

~~You're connecting `User`s and `Bot`s through the same URL ([why is this important?](https://discord.com/developers/docs/topics/gateway#connecting-to-the-gateway))~~

Oops haha, I was wrong about that last part. Even still, connecting to a decommissioned API socket will either
> a) not work at all
> b) be a dead-giveaway that you're doing stuff you aren't supposed to be with the API (similar to bad `"IDENTIFY"` payloads)

### So... Anything Else?
My last pieces of advice? Learn from the past, and try to reproduce [the most secure selfbot frameworks](https://github.com/13-05/hidden-disc-docs#frameworks), as well as to be careful with so-called ["risky actions"](https://gist.github.com/13-05/c7691d633fd011e96aceebe889a5a4a9) by other user-api libraries, and also make sure that you do all testing of something to this scale on an alternate account.

Other than that... Have fun, good luck, and always remember that it's never illegal unless you get caught! ;)

TL;DR: Good idea, bad implementation currently and may easily flag accounts, as well as not even work in some cases.

**Comment by tungdo0602** [comment]
ok i'll try

**Comment by 13-05** [comment]
👍

**Comment by 13-05** [comment]
just tested the beta and current versions, it seems you don't need to bypass cors. my mistake haha.

**Comment by tungdo0602** [comment]
ok lol
